File: src/main/java/org/cubectl/identity/user/UserRepository.java
package org.cubectl.identity.user;

import java.util.Optional;
import java.util.UUID;

import org.springframework.data.jpa.repository.EntityGraph;
import org.springframework.data.jpa.repository.JpaRepository;

public interface UserRepository extends JpaRepository<UserEntity, UUID> {

    boolean existsByUsername(String username);

    Optional<UserEntity> findByUsername(String username);

    @EntityGraph(attributePaths = {"roles", "roles.permissions"})
    Optional<UserEntity> findWithRolesById(UUID id);

    @EntityGraph(attributePaths = {"roles", "roles.permissions"})
    Optional<UserEntity> findWithRolesByUsername(String username);
}

File: src/main/kotlin/org/cubectl/identity/user/UserRepository.kt
package org.cubectl.identity.user

import org.springframework.data.jpa.repository.EntityGraph
import org.springframework.data.jpa.repository.JpaRepository
import java.util.Optional
import java.util.UUID

interface UserRepository : JpaRepository<UserEntity, UUID> {

    fun existsByUsername(username: String): Boolean

    @EntityGraph(attributePaths = ["roles", "roles.permissions"])
    fun findWithRolesByUsername(username: String): Optional<UserEntity>

    @EntityGraph(attributePaths = ["roles", "roles.permissions"])
    fun findWithRolesById(id: UUID): Optional<UserEntity>
}
